### PR TITLE
Support customisation of NpgsqlDataSourceBuilder during component registration

### LINKED ISSUE #1479: Support modifying the NpgsqlDataSourceBuilder during component registration
The `NpgsqlDataSourceBuilder` allows you to do modifications to things such as the type mappings when working with PostgreSQL, a use-case that I'm looking at is [pgvector](https://github.com/pgvector/pgvector-dotnet) to support Embeddings in AI projects.

But the current component registration does not provide a hook to work with the `NpgsqlDataSourceBuilder`, meaning it's not possible to do this.

### PR DESCRIPTION
**Pull request by aaronpowell**
Added optional callback to component registration that allows you to modify the DataSourceBuilder

Fixes #1479
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1481)

### CODE REVIEW

File: src/Components/Aspire.Npgsql/AspirePostgreSqlNpgsqlExtensions.cs [hunk 1]
```diff
@@ -99,14 +101,19 @@ public static void AddKeyedNpgsqlDataSource(this IHostApplicationBuilder builder
...
         {
             // delay validating the ConnectionString until the DataSource is requested. This ensures an exception doesn't happen until a Logger is established.
             builder.Services.AddNpgsqlDataSource(settings.ConnectionString ?? string.Empty, dataSourceBuilder =>
             {
                 ValidateConnection();
+
+                if (configureDataSourceBuilder is not null)
+                {
+                    configureDataSourceBuilder(dataSourceBuilder);
+                }
```
Comment (sebastienros) [inline src/Components/Aspire.Npgsql/AspirePostgreSqlNpgsqlExtensions.cs:116]:
```suggestion
                configureDataSourceBuilder?.Invoke(dataSourceBuilder);
```

nit: I believe people that syntax in this repos

File: src/Components/Aspire.Npgsql/AspirePostgreSqlNpgsqlExtensions.cs [hunk 2]
```diff
@@ -119,6 +126,12 @@ private static void RegisterNpgsqlServices(this IHostApplicationBuilder builder,
 
                 var dataSourceBuilder = new NpgsqlDataSourceBuilder(settings.ConnectionString);
                 dataSourceBuilder.UseLoggerFactory(serviceProvider.GetService<ILoggerFactory>());
+
+                if (configureDataSourceBuilder is not null)
+                {
+                    configureDataSourceBuilder(dataSourceBuilder);
+                }
```
Comment (sebastienros) [inline src/Components/Aspire.Npgsql/AspirePostgreSqlNpgsqlExtensions.cs:133]:
```suggestion
                configureDataSourceBuilder?.Invoke(dataSourceBuilder);
```

**Comment by aaronpowell** [comment]
Good catch @sebastienros - should've looked at the coding conventions more

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**